### DISCUSSION TITLE
ESP MicroPython Flashing / Setup Utility

### DISCUSSION THREAD

**Discussion by sethstenzel** [micropython/micropython#11027] [Show and tell]
Greetings everyone,

I'm sharing a small utility I wrote for work. It is a tool (currently windows only) I built for work that allowed me to pass updated firmware and zipped python package files around to others that we didn't want to have to install anything extra when working with a POC device. My work version is for a specific board, so It does not have the other chips, baud, and offset dropdowns; however, I have added them in this public version for more utility. If you have a specific board you are pushing things to, you could do something similar, I have only tested it against a select number of ESP boards, so if you come across an issue, open a GL issue, and I'll try and get a copy of your board and resolve.

Written in python, using PySimpleGUI and PyInstaller. There may be better options, but I have yet to encounter them. You will get the blue warning window the first time you open it. Alternatively, you can build with AutoPyToExe after tweaking the `pyinstaller.json` file to match your paths..

While this does not have advanced options you might need from esptools, for me, it covers 99% of what I'm doing with different boards and MicroPython.

You can use it to only flash firmware, only flash a zip archive with your python files or both at the same time.

![alt text](https://github.com/seth-c-stenzel/ESP-MicroPython-Setup-Utility/blob/main/screenshots/ESP%20MicroPython%20Setup%20Utility.jpg?raw=true)

**EDIT**
Forgot the URL to the repo, thanks @sosi-deadeye.
[https://github.com/seth-c-stenzel/ESP-MicroPython-Setup-Utility](https://github.com/seth-c-stenzel/ESP-MicroPython-Setup-Utility)

**Comment by sosi-deadeye** [comment]
You have forgotten to put a link to your repository: https://github.com/seth-c-stenzel/ESP-MicroPython-Setup-Utility

**Comment by walklinewang** [comment]
I have tried with esp32c3 module, erase and flash worked fine, but uploading .py sourcecode(a .zip file contains 3 files in 2 folders) stucked with no reason.

**Comment by sethstenzel** [comment]
Thanks for the heads up! Could you link the board you used? I'll see if I
can figure it out once I have one.

**Reply by walklinewang** [reply]
Hey pal, I can't name it or give you a link because of the board was designed by myself, and just used an esp32c3 module on it, but you can find out some boards like `Seeed XIAO ESP32C3` that @robert-hh used to help me to solve my problem in issue https://github.com/micropython/micropython/issues/10980

**Reply by sethstenzel** [reply]
I've ordered one of those `Seeed XIAO ESP32C3` modules, I think I know what the issue is, though. The app does a hard reset, and I believe that those boards, when hard reset actually reset the serial connection, the pc briefly sees them disconnecting and reconnecting. Once it gets here, I'll test it. If I'm right, would be an easy fix.

**Reply by sethstenzel** [reply]
I updated the repo. My guess was correct, in order for you to upload the files you need to reset the device after flash, or you could unplug and plug it back in. The esptool does not seem able to reset that device itself. For that module, I have added a warning so that when the time to unplug/reset has come, you can do so and then simply press the reset button on device and hit enter to continue with the upload.

```
################################################################################
WARNING: esp32c3 chip detected, unable to force reboot, but reboot/reset required!

Please press the reset button, or unplug and replug device and then
hit ENTER to continue with software upload.
```
